src/core/Akka/IO/SocketEventArgsPool.cs

```diff
@@ -21,48 +22,81 @@ public interface ISocketEventArgsPool
     {
         SocketAsyncEventArgs Acquire(IActorRef actor);
         void Release(SocketAsyncEventArgs e);
+        
+        BufferPoolInfo BufferPoolInfo { get; }
     }
 
     internal class PreallocatedSocketEventAgrsPool : ISocketEventArgsPool
     {
+        private readonly IBufferPool _bufferPool;
         private readonly EventHandler<SocketAsyncEventArgs> _onComplete;
-        private readonly ConcurrentStack<SocketAsyncEventArgs> _pool = new ConcurrentStack<SocketAsyncEventArgs>();
+        private readonly ConcurrentQueue<SocketAsyncEventArgs> _pool = new ConcurrentQueue<SocketAsyncEventArgs>();
 
-        public PreallocatedSocketEventAgrsPool(int initSize, EventHandler<SocketAsyncEventArgs> onComplete)
+        
+        public PreallocatedSocketEventAgrsPool(int initSize, IBufferPool bufferPool, EventHandler<SocketAsyncEventArgs> onComplete)
         {
+            _bufferPool = bufferPool;
             _onComplete = onComplete;
             for (var i = 0; i < initSize; i++)
             {
-                var e = CreateSocketAsyncEventArgs();
-                _pool.Push(e);
+                var e = new SocketAsyncEventArgs { UserToken = null };
+                e.Completed += _onComplete;
+                _pool.Enqueue(e);
             }
         }
 
         public SocketAsyncEventArgs Acquire(IActorRef actor)
         {
-            if (!_pool.TryPop(out var e))
-                e = CreateSocketAsyncEventArgs();
+            var buffer = _bufferPool.Rent();
+            var acquired = false;
+            SocketAsyncEventArgs e = null;
+            while (!acquired)
+            {
+                try
+                {
+                    if (!_pool.TryDequeue(out e))
+                        e = new SocketAsyncEventArgs();
 
-            e.UserToken = actor;
+                    e.SetBuffer(buffer.Array, buffer.Offset, buffer.Count);
```

Comment:
This is fine because we only rent once and if we fail to acquire an instance, we immediately dispose it.